File: LEETCODE/LEETCODE-JAVA/21.java
//21. Merge Two Sorted Lists
//https://leetcode.com/problems/merge-two-sorted-lists/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode start=new ListNode(), res=start;
        
        if(l1==null)
            return l2;
        if(l2==null)
            return l1;
        
        while(l1!=null && l2!=null){
            if(l1.val<l2.val){
                res.next=l1;
                l1=l1.next;
            }
            else{
                res.next=l2;
                l2=l2.next;
            }
            
            res=res.next;
        }
        
        if(l1!=null)
            res.next=l1;
        if(l2!=null)
            res.next=l2;
        
        return start.next;
    }
}

File: LEETCODE/LEETCODE-PYTHON/21.py
# 21. Merge Two Sorted Lists
# https://leetcode.com/problems/merge-two-sorted-lists/description/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

# Method-1 using recursion(TC-O(m+n),   SC-O(m+n))
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        if list1 is None:
            return list2
        elif list2 is None:
            return list1
        elif list1.val<list2.val:
            list1.next=self.mergeTwoLists(list1.next,list2)
            return list1
        else:
            list2.next=self.mergeTwoLists(list2.next,list1)
            return list2


# Method-2 using iteration (TC-O(m+n),   SC-O(1))
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        prehead=ListNode(-1)

        prev=prehead

        while list1 and list2:
            if list1.val<list2.val:
                prev.next=list1
                list1=list1.next
            else:
                prev.next=list2
                list2=list2.next

            prev=prev.next

        prev.next = list1 or list2

        return prehead.next